File: crates/headgate-migrate/src/postgres.rs
use std::collections::HashSet;

use headgate_sql::{PostgresNamespace, quote_identifier};
use tokio_postgres::{Client, GenericClient};

use crate::schema::{POSTGRES_COLUMNS, POSTGRES_INDEXES, POSTGRES_TRIGGERS, STATES, TABLES};
use crate::{
    AppliedMigration, Backend, Direction, InstallationState, MigrateOptions, MigrateResult,
    MigrationError, checksum, latest_version, migrations, plan,
};

const CREATE_HISTORY: &str = "
CREATE TABLE IF NOT EXISTS headgate_schema_migration (
  line          text   NOT NULL DEFAULT 'main',
  version       bigint NOT NULL,
  name          text   NOT NULL,
  checksum      text   NOT NULL,
  applied_at_ms bigint NOT NULL,
  PRIMARY KEY (line, version)
)";

#[derive(Clone, Debug)]
pub struct PostgresValidation {
    pub state: InstallationState,
    pub current_version: u32,
    pub latest_version: u32,
    pub applied: Vec<AppliedMigration>,
    pub messages: Vec<String>,
}

impl PostgresValidation {
    pub fn is_ok(&self) -> bool {
        self.messages.is_empty()
    }
}

async fn namespace_for(
    client: &Client,
    schema: Option<&str>,
) -> Result<PostgresNamespace, MigrationError> {
    let name = match schema {
        Some(name) => name.to_owned(),
        None => client
            .query_one("SELECT current_schema()", &[])
            .await?
            .get(0),
    };
    let namespace = PostgresNamespace::explicit(&name).map_err(MigrationError::Invalid)?;
    let exists: bool = client
        .query_one(
            "SELECT EXISTS (SELECT 1 FROM pg_namespace WHERE nspname = $1)",
            &[&name],
        )
        .await?
        .get(0);
    if !exists {
        return Err(MigrationError::Invalid(format!(
            "Postgres schema {name:?} does not exist; create it before migrating"
        )));
    }
    Ok(namespace)
}

fn relation_name(namespace: &PostgresNamespace, relation: &str) -> String {
    format!(
        "{}.{}",
        quote_identifier(namespace.name().expect("resolved namespace")),
        quote_identifier(relation)
    )
}

async fn relation_exists<C>(
    client: &C,
    namespace: &PostgresNamespace,
    relation: &str,
) -> Result<bool, MigrationError>
where
    C: GenericClient + Sync,
{
    let relation = relation_name(namespace, relation);
    Ok(client
        .query_one("SELECT to_regclass($1) IS NOT NULL", &[&relation])
        .await?
        .get(0))
}

async fn read_history<C>(
    client: &C,
    namespace: &PostgresNamespace,
) -> Result<Vec<AppliedMigration>, MigrationError>
where
    C: GenericClient + Sync,
{
    let sql = namespace.render(
        "SELECT version, name, checksum, applied_at_ms
           FROM headgate_schema_migration
          WHERE line = 'main'
          ORDER BY version",
    );
    let rows = client.query(sql.as_ref(), &[]).await?;
    rows.into_iter()
        .map(|row| {
            let version: i64 = row.get(0);
            let version = u32::try_from(version).map_err(|_| {
                MigrationError::History(format!("version {version} is outside u32"))
            })?;
            Ok(AppliedMigration {
                version,
                name: row.get(1),
                checksum: row.get(2),
                applied_at_ms: row.get(3),
            })
        })
        .collect()
}

async fn applied_postgres_scoped(
    client: &Client,
    namespace: &PostgresNamespace,
) -> Result<(InstallationState, Vec<AppliedMigration>), MigrationError> {
    let has_history = relation_exists(client, namespace, "headgate_schema_migration").await?;
    let has_schema = relation_exists(client, namespace, "headgate_job").await?;
    if !has_history {
        return Ok((
            if has_schema {
                InstallationState::Unversioned
            } else {
                InstallationState::Empty
            },
            Vec::new(),
        ));
    }
    let applied = read_history(client, namespace).await?;
    let state = if applied.is_empty() && !has_schema {
        InstallationState::Empty
    } else if applied.is_empty() && has_schema {
        InstallationState::Unversioned
    } else {
        InstallationState::Versioned
    };
    Ok((state, applied))
}

pub async fn applied_postgres(
    client: &Client,
) -> Result<(InstallationState, Vec<AppliedMigration>), MigrationError> {
    let namespace = namespace_for(client, None).await?;
    applied_postgres_scoped(client, &namespace).await
}

pub async fn applied_postgres_in_schema(
    client: &Client,
    schema: &str,
) -> Result<(InstallationState, Vec<AppliedMigration>), MigrationError> {
    let namespace = namespace_for(client, Some(schema)).await?;
    applied_postgres_scoped(client, &namespace).await
}

async fn missing_schema<C>(
    client: &C,
    namespace: &PostgresNamespace,
) -> Result<Vec<String>, MigrationError>
where
    C: GenericClient + Sync,
{
    let schema = namespace.name().expect("resolved namespace");
    let mut missing = Vec::new();
    let tables: HashSet<String> = client
        .query(
            "SELECT table_name
               FROM information_schema.tables
              WHERE table_schema = $1 AND table_name LIKE 'headgate_%'",
            &[&schema],
        )
        .await?
        .into_iter()
        .map(|row| row.get(0))
        .collect();
    for table in TABLES {
        if !tables.contains(*table) {
            missing.push(format!("missing table {table}"));
        }
    }

    let columns: HashSet<(String, String)> = client
        .query(
            "SELECT table_name, column_name
               FROM information_schema.columns
              WHERE table_schema = $1 AND table_name LIKE 'headgate_%'",
            &[&schema],
        )
        .await?
        .into_iter()
        .map(|row| (row.get(0), row.get(1)))
        .collect();
    for required in POSTGRES_COLUMNS {
        if !columns.contains(&(required.table.to_owned(), required.column.to_owned())) {
            missing.push(format!(
                "missing column {}.{}",
                required.table, required.column
            ));
        }
    }

    let indexes: HashSet<String> = client
        .query(
            "SELECT indexname FROM pg_indexes
              WHERE schemaname = $1 AND indexname LIKE 'headgate_%'",
            &[&schema],
        )
        .await?
        .into_iter()
        .map(|row| row.get(0))
        .collect();
    for index in POSTGRES_INDEXES {
        if !indexes.contains(*index) {
            missing.push(format!("missing index {index}"));
        }
    }

    let triggers: HashSet<String> = client
        .query(
            "SELECT t.tgname
               FROM pg_trigger t
               JOIN pg_class c ON c.oid = t.tgrelid
               JOIN pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = $1 AND NOT t.tgisinternal",
            &[&schema],
        )
        .await?
        .into_iter()
        .map(|row| row.get(0))
        .collect();
    for trigger in POSTGRES_TRIGGERS {
        if !triggers.contains(*trigger) {
            missing.push(format!("missing trigger {trigger}"));
        }
    }

    let states: Vec<String> = client
        .query(
            "SELECT e.enumlabel
               FROM pg_type t
               JOIN pg_namespace n ON n.oid = t.typnamespace
               JOIN pg_enum e ON e.enumtypid = t.oid
              WHERE n.nspname = $1 AND t.typname = 'headgate_state'
              ORDER BY e.enumsortorder",
            &[&schema],
        )
        .await?
        .into_iter()
        .map(|row| row.get(0))
        .collect();
    let expected_states: Vec<String> = STATES.iter().map(|state| (*state).to_owned()).collect();
    if states != expected_states {
        missing.push(format!(
            "headgate_state labels are {:?}, expected {:?}",
            states, expected_states
        ));
    }
    Ok(missing)
}

async fn validate_postgres_scoped(
    client: &Client,
    namespace: &PostgresNamespace,
) -> Result<PostgresValidation, MigrationError> {
    let (state, applied) = applied_postgres_scoped(client, namespace).await?;
    let current_version = applied.last().map_or(0, |row| row.version);
    let latest = latest_version(Backend::Postgres);
    let mut messages = Vec::new();
    match state {
        InstallationState::Empty => messages.push("headgate schema is not installed".into()),
        InstallationState::Unversioned => {
            messages.push("headgate schema exists without migration history".into());
            messages.extend(missing_schema(client, namespace).await?);
        }
        InstallationState::Versioned => {
            if let Err(error) = crate::validate_history(Backend::Postgres, &applied) {
                messages.push(error.to_string());
            }
            if current_version != latest {
                messages.push(format!(
                    "schema is at version {current_version}, embedded latest is {latest}"
                ));
            } else {
                messages.extend(missing_schema(client, namespace).await?);
            }
        }
    }
    Ok(PostgresValidation {
        state,
        current_version,
        latest_version: latest,
        applied,
        messages,
    })
}

pub async fn validate_postgres(client: &Client) -> Result<PostgresValidation, MigrationError> {
    let namespace = namespace_for(client, None).await?;
    validate_postgres_scoped(client, &namespace).await
}

pub async fn validate_postgres_in_schema(
    client: &Client,
    schema: &str,
) -> Result<PostgresValidation, MigrationError> {
    let namespace = namespace_for(client, Some(schema)).await?;
    validate_postgres_scoped(client, &namespace).await
}

async fn migrate_postgres_scoped(
    client: &mut Client,
    namespace: &PostgresNamespace,
    direction: Direction,
    options: MigrateOptions,
) -> Result<MigrateResult, MigrationError> {
    let (state, applied) = applied_postgres_scoped(client, namespace).await?;
    if state == InstallationState::Unversioned {
        return Err(MigrationError::UnversionedSchema);
    }
    let planned = plan(Backend::Postgres, &applied, direction, options)?;
    if options.dry_run {
        return Ok(MigrateResult {
            dry_run: true,
            steps: planned,
        });
    }
    let create_history = namespace.render(CREATE_HISTORY);
    client.batch_execute(create_history.as_ref()).await?;
    let mut executed = Vec::new();
    for intended in planned {
        let tx = client.transaction().await?;
        let lock =
            namespace.render("LOCK TABLE headgate_schema_migration IN ACCESS EXCLUSIVE MODE");
        tx.batch_execute(lock.as_ref()).await?;
        let live = read_history(&tx, namespace).await?;
        crate::validate_history(Backend::Postgres, &live)?;
        let current = live.last().map_or(0, |row| row.version);
        match direction {
            Direction::Up => {
                if current >= intended.migration.version {
                    tx.commit().await?;
                    continue;
                }
                if current + 1 != intended.migration.version {
                    return Err(MigrationError::History(format!(
                        "concurrent migration moved current version to {current}; expected {}",
                        intended.migration.version - 1
                    )));
                }
                let up_sql = namespace.render(intended.migration.up_sql);
                tx.batch_execute(up_sql.as_ref()).await?;
                // The manifest describes LATEST, not every historical intermediate.
                // Validate it only after the step that is supposed to produce latest;
                // otherwise adding v2 makes a fresh v1 transaction fail for lacking v2.
                if intended.migration.version == latest_version(Backend::Postgres) {
                    let schema_errors = missing_schema(&tx, namespace).await?;
                    if !schema_errors.is_empty() {
                        return Err(MigrationError::schema(schema_errors));
                    }
                }
                let version = i64::from(intended.migration.version);
                let migration_checksum = checksum(intended.migration);
                let insert = namespace.render(
                    "INSERT INTO headgate_schema_migration
                       (line, version, name, checksum, applied_at_ms)
                     VALUES ('main', $1, $2, $3,
                             (EXTRACT(EPOCH FROM clock_timestamp()) * 1000)::bigint)",
                );
                tx.execute(
                    insert.as_ref(),
                    &[&version, &intended.migration.name, &migration_checksum],
                )
                .await?;
            }
            Direction::Down => {
                if current < intended.migration.version {
                    tx.commit().await?;
                    continue;
                }
                if current != intended.migration.version {
                    return Err(MigrationError::History(format!(
                        "cannot migrate down version {}; current version is {current}",
                        intended.migration.version
                    )));
                }
                let down_sql = namespace.render(intended.migration.down_sql);
                tx.batch_execute(down_sql.as_ref()).await?;
                let version = i64::from(intended.migration.version);
                let delete = namespace.render(
                    "DELETE FROM headgate_schema_migration
                      WHERE line = 'main' AND version = $1",
                );
                tx.execute(delete.as_ref(), &[&version]).await?;
            }
        }
        tx.commit().await?;
        executed.push(intended);
    }
    Ok(MigrateResult {
        dry_run: false,
        steps: executed,
    })
}

pub async fn migrate_postgres(
    client: &mut Client,
    direction: Direction,
    options: MigrateOptions,
) -> Result<MigrateResult, MigrationError> {
    let namespace = namespace_for(client, None).await?;
    migrate_postgres_scoped(client, &namespace, direction, options).await
}

pub async fn migrate_postgres_in_schema(
    client: &mut Client,
    schema: &str,
    direction: Direction,
    options: MigrateOptions,
) -> Result<MigrateResult, MigrationError> {
    let namespace = namespace_for(client, Some(schema)).await?;
    migrate_postgres_scoped(client, &namespace, direction, options).await
}

async fn adopt_postgres_scoped(
    client: &mut Client,
    namespace: &PostgresNamespace,
) -> Result<Vec<AppliedMigration>, MigrationError> {
    let (state, existing) = applied_postgres_scoped(client, namespace).await?;
    if state == InstallationState::Versioned {
        crate::validate_history(Backend::Postgres, &existing)?;
        return Ok(existing);
    }
    if state == InstallationState::Empty {
        return Err(MigrationError::Invalid(
            "cannot adopt an empty database; migrate up instead".into(),
        ));
    }
    let create_history = namespace.render(CREATE_HISTORY);
    client.batch_execute(create_history.as_ref()).await?;
    let tx = client.transaction().await?;
    let lock = namespace.render("LOCK TABLE headgate_schema_migration IN ACCESS EXCLUSIVE MODE");
    tx.batch_execute(lock.as_ref()).await?;
    let live = read_history(&tx, namespace).await?;
    if !live.is_empty() {
        crate::validate_history(Backend::Postgres, &live)?;
        tx.commit().await?;
        return Ok(live);
    }
    let schema_errors = missing_schema(&tx, namespace).await?;
    if !schema_errors.is_empty() {
        return Err(MigrationError::schema(schema_errors));
    }
    let insert = namespace.render(
        "INSERT INTO headgate_schema_migration
           (line, version, name, checksum, applied_at_ms)
         VALUES ('main', $1, $2, $3,
                 (EXTRACT(EPOCH FROM clock_timestamp()) * 1000)::bigint)",
    );
    for migration in migrations(Backend::Postgres) {
        let version = i64::from(migration.version);
        let migration_checksum = checksum(migration);
        tx.execute(
            insert.as_ref(),
            &[&version, &migration.name, &migration_checksum],
        )
        .await?;
    }
    tx.commit().await?;
    applied_postgres_scoped(client, namespace)
        .await
        .map(|(_, applied)| applied)
}

pub async fn adopt_postgres(client: &mut Client) -> Result<Vec<AppliedMigration>, MigrationError> {
    let namespace = namespace_for(client, None).await?;
    adopt_postgres_scoped(client, &namespace).await
}

pub async fn adopt_postgres_in_schema(
    client: &mut Client,
    schema: &str,
) -> Result<Vec<AppliedMigration>, MigrationError> {
    let namespace = namespace_for(client, Some(schema)).await?;
    adopt_postgres_scoped(client, &namespace).await
}

File: go/headgatemigrate/postgres.go
package headgatemigrate

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5"
	"github.com/mujhtech/headgate/go/postgressql"
)

const createPostgresHistory = `
CREATE TABLE IF NOT EXISTS headgate_schema_migration (
  line          text   NOT NULL DEFAULT 'main',
  version       bigint NOT NULL,
  name          text   NOT NULL,
  checksum      text   NOT NULL,
  applied_at_ms bigint NOT NULL,
  PRIMARY KEY (line, version)
)`

type PostgresValidation struct {
	State          InstallationState
	CurrentVersion int
	LatestVersion  int
	Applied        []AppliedMigration
	Messages       []string
}

func (v PostgresValidation) OK() bool { return len(v.Messages) == 0 }

type pgReader interface {
	Query(context.Context, string, ...any) (pgx.Rows, error)
	QueryRow(context.Context, string, ...any) pgx.Row
}

func postgresNamespace(
	ctx context.Context,
	conn *pgx.Conn,
	explicitSchema *string,
) (postgressql.Namespace, error) {
	var name string
	if explicitSchema == nil {
		if err := conn.QueryRow(ctx, "SELECT current_schema()").Scan(&name); err != nil {
			return postgressql.Namespace{}, err
		}
	} else {
		name = *explicitSchema
	}
	namespace, err := postgressql.NewNamespace(name)
	if err != nil {
		return postgressql.Namespace{}, err
	}
	var exists bool
	if err := conn.QueryRow(ctx,
		"SELECT EXISTS (SELECT 1 FROM pg_namespace WHERE nspname = $1)", name,
	).Scan(&exists); err != nil {
		return postgressql.Namespace{}, err
	}
	if !exists {
		return postgressql.Namespace{}, fmt.Errorf(
			"postgres schema %q does not exist; create it before migrating", name,
		)
	}
	return namespace, nil
}

func pgRelationExists(
	ctx context.Context,
	db pgReader,
	namespace postgressql.Namespace,
	relation string,
) (bool, error) {
	var exists bool
	err := db.QueryRow(ctx, "SELECT to_regclass($1) IS NOT NULL", namespace.Qualified(relation)).Scan(&exists)
	return exists, err
}

func readPostgresHistory(
	ctx context.Context,
	db pgReader,
	namespace postgressql.Namespace,
) ([]AppliedMigration, error) {
	rows, err := db.Query(ctx, namespace.Render(`
SELECT version, name, checksum, applied_at_ms
  FROM headgate_schema_migration
 WHERE line = 'main'
 ORDER BY version`))
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	result := make([]AppliedMigration, 0)
	for rows.Next() {
		var row AppliedMigration
		if err := rows.Scan(&row.Version, &row.Name, &row.Checksum, &row.AppliedAtMS); err != nil {
			return nil, err
		}
		if row.Version < 0 {
			return nil, &HistoryError{Message: fmt.Sprintf("version %d is negative", row.Version)}
		}
		result = append(result, row)
	}
	return result, rows.Err()
}

func appliedPostgresScoped(
	ctx context.Context,
	conn *pgx.Conn,
	namespace postgressql.Namespace,
) (InstallationState, []AppliedMigration, error) {
	hasHistory, err := pgRelationExists(ctx, conn, namespace, "headgate_schema_migration")
	if err != nil {
		return "", nil, err
	}
	hasSchema, err := pgRelationExists(ctx, conn, namespace, "headgate_job")
	if err != nil {
		return "", nil, err
	}
	if !hasHistory {
		if hasSchema {
			return Unversioned, nil, nil
		}
		return Empty, nil, nil
	}
	applied, err := readPostgresHistory(ctx, conn, namespace)
	if err != nil {
		return "", nil, err
	}
	if len(applied) == 0 {
		if hasSchema {
			return Unversioned, nil, nil
		}
		return Empty, nil, nil
	}
	return Versioned, applied, nil
}

func AppliedPostgres(ctx context.Context, conn *pgx.Conn) (InstallationState, []AppliedMigration, error) {
	namespace, err := postgresNamespace(ctx, conn, nil)
	if err != nil {
		return "", nil, err
	}
	return appliedPostgresScoped(ctx, conn, namespace)
}

func AppliedPostgresInSchema(
	ctx context.Context,
	conn *pgx.Conn,
	schema string,
) (InstallationState, []AppliedMigration, error) {
	namespace, err := postgresNamespace(ctx, conn, &schema)
	if err != nil {
		return "", nil, err
	}
	return appliedPostgresScoped(ctx, conn, namespace)
}

func missingPostgresSchema(
	ctx context.Context,
	db pgReader,
	namespace postgressql.Namespace,
) ([]string, error) {
	missing := make([]string, 0)
	tables := map[string]bool{}
	rows, err := db.Query(ctx, `
SELECT table_name
  FROM information_schema.tables
 WHERE table_schema = $1 AND table_name LIKE 'headgate_%'`, namespace.Name())
	if err != nil {
		return nil, err
	}
	for rows.Next() {
		var name string
		if err := rows.Scan(&name); err != nil {
			rows.Close()
			return nil, err
		}
		tables[name] = true
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return nil, err
	}
	rows.Close()
	for _, table := range requiredTables {
		if !tables[table] {
			missing = append(missing, "missing table "+table)
		}
	}

	columns := map[string]bool{}
	rows, err = db.Query(ctx, `
SELECT table_name, column_name
  FROM information_schema.columns
 WHERE table_schema = $1 AND table_name LIKE 'headgate_%'`, namespace.Name())
	if err != nil {
		return nil, err
	}
	for rows.Next() {
		var table, column string
		if err := rows.Scan(&table, &column); err != nil {
			rows.Close()
			return nil, err
		}
		columns[table+"."+column] = true
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return nil, err
	}
	rows.Close()
	for _, column := range postgresJobColumns {
		if !columns["headgate_job."+column] {
			missing = append(missing, "missing column headgate_job."+column)
		}
	}
	for table, list := range commonColumns {
		for _, column := range list {
			if !columns[table+"."+column] {
				missing = append(missing, "missing column "+table+"."+column)
			}
		}
	}
	for _, column := range []string{"key", "at_ms"} {
		if !columns["headgate_effect."+column] {
			missing = append(missing, "missing column headgate_effect."+column)
		}
	}

	indexes := map[string]bool{}
	rows, err = db.Query(ctx, `
SELECT indexname FROM pg_indexes
 WHERE schemaname = $1 AND indexname LIKE 'headgate_%'`, namespace.Name())
	if err != nil {
		return nil, err
	}
	for rows.Next() {
		var name string
		if err := rows.Scan(&name); err != nil {
			rows.Close()
			return nil, err
		}
		indexes[name] = true
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return nil, err
	}
	rows.Close()
	for _, index := range postgresIndexes {
		if !indexes[index] {
			missing = append(missing, "missing index "+index)
		}
	}

	triggers := map[string]bool{}
	rows, err = db.Query(ctx, `
SELECT t.tgname
  FROM pg_trigger t
  JOIN pg_class c ON c.oid = t.tgrelid
  JOIN pg_namespace n ON n.oid = c.relnamespace
 WHERE n.nspname = $1 AND NOT t.tgisinternal`, namespace.Name())
	if err != nil {
		return nil, err
	}
	for rows.Next() {
		var name string
		if err := rows.Scan(&name); err != nil {
			rows.Close()
			return nil, err
		}
		triggers[name] = true
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return nil, err
	}
	rows.Close()
	for _, trigger := range postgresTriggers {
		if !triggers[trigger] {
			missing = append(missing, "missing trigger "+trigger)
		}
	}

	rows, err = db.Query(ctx, `
SELECT e.enumlabel
  FROM pg_type t
  JOIN pg_namespace n ON n.oid = t.typnamespace
  JOIN pg_enum e ON e.enumtypid = t.oid
 WHERE n.nspname = $1 AND t.typname = 'headgate_state'
 ORDER BY e.enumsortorder`, namespace.Name())
	if err != nil {
		return nil, err
	}
	states := make([]string, 0)
	for rows.Next() {
		var state string
		if err := rows.Scan(&state); err != nil {
			rows.Close()
			return nil, err
		}
		states = append(states, state)
	}
	if err := rows.Err(); err != nil {
		rows.Close()
		return nil, err
	}
	rows.Close()
	if !equalStrings(states, stateLabels) {
		missing = append(missing, fmt.Sprintf("headgate_state labels are %v, expected %v", states, stateLabels))
	}
	return missing, nil
}

func validatePostgresScoped(
	ctx context.Context,
	conn *pgx.Conn,
	namespace postgressql.Namespace,
) (PostgresValidation, error) {
	state, applied, err := appliedPostgresScoped(ctx, conn, namespace)
	if err != nil {
		return PostgresValidation{}, err
	}
	current := 0
	if len(applied) != 0 {
		current = applied[len(applied)-1].Version
	}
	latest := LatestVersion(Postgres)
	messages := make([]string, 0)
	switch state {
	case Empty:
		messages = append(messages, "headgate schema is not installed")
	case Unversioned:
		messages = append(messages, "headgate schema exists without migration history")
		shape, err := missingPostgresSchema(ctx, conn, namespace)
		if err != nil {
			return PostgresValidation{}, err
		}
		messages = append(messages, shape...)
	case Versioned:
		if err := ValidateHistory(Postgres, applied); err != nil {
			messages = append(messages, err.Error())
		}
		if current != latest {
			messages = append(messages, fmt.Sprintf("schema is at version %d, embedded latest is %d", current, latest))
		} else {
			shape, err := missingPostgresSchema(ctx, conn, namespace)
			if err != nil {
				return PostgresValidation{}, err
			}
			messages = append(messages, shape...)
		}
	}
	return PostgresValidation{
		State: state, CurrentVersion: current, LatestVersion: latest,
		Applied: applied, Messages: messages,
	}, nil
}

func ValidatePostgres(ctx context.Context, conn *pgx.Conn) (PostgresValidation, error) {
	namespace, err := postgresNamespace(ctx, conn, nil)
	if err != nil {
		return PostgresValidation{}, err
	}
	return validatePostgresScoped(ctx, conn, namespace)
}

func ValidatePostgresInSchema(
	ctx context.Context,
	conn *pgx.Conn,
	schema string,
) (PostgresValidation, error) {
	namespace, err := postgresNamespace(ctx, conn, &schema)
	if err != nil {
		return PostgresValidation{}, err
	}
	return validatePostgresScoped(ctx, conn, namespace)
}

func migratePostgresScoped(
	ctx context.Context,
	conn *pgx.Conn,
	namespace postgressql.Namespace,
	direction Direction,
	options Options,
) (Result, error) {
	state, applied, err := appliedPostgresScoped(ctx, conn, namespace)
	if err != nil {
		return Result{}, err
	}
	if state == Unversioned {
		return Result{}, ErrUnversionedSchema
	}
	planned, err := Plan(Postgres, applied, direction, options)
	if err != nil {
		return Result{}, err
	}
	if options.DryRun {
		return Result{DryRun: true, Steps: planned}, nil
	}
	if _, err := conn.Exec(ctx, namespace.Render(createPostgresHistory)); err != nil {
		return Result{}, err
	}
	executed := make([]Step, 0, len(planned))
	for _, intended := range planned {
		didRun, err := applyPostgresStep(ctx, conn, namespace, direction, intended)
		if err != nil {
			return Result{}, err
		}
		if didRun {
			executed = append(executed, intended)
		}
	}
	return Result{Steps: executed}, nil
}

func MigratePostgres(
	ctx context.Context,
	conn *pgx.Conn,
	direction Direction,
	options Options,
) (Result, error) {
	namespace, err := postgresNamespace(ctx, conn, nil)
	if err != nil {
		return Result{}, err
	}
	return migratePostgresScoped(ctx, conn, namespace, direction, options)
}

func MigratePostgresInSchema(
	ctx context.Context,
	conn *pgx.Conn,
	schema string,
	direction Direction,
	options Options,
) (Result, error) {
	namespace, err := postgresNamespace(ctx, conn, &schema)
	if err != nil {
		return Result{}, err
	}
	return migratePostgresScoped(ctx, conn, namespace, direction, options)
}

func applyPostgresStep(
	ctx context.Context,
	conn *pgx.Conn,
	namespace postgressql.Namespace,
	direction Direction,
	intended Step,
) (bool, error) {
	tx, err := conn.Begin(ctx)
	if err != nil {
		return false, err
	}
	defer func() { _ = tx.Rollback(ctx) }()
	if _, err := tx.Exec(ctx, namespace.Render(
		"LOCK TABLE headgate_schema_migration IN ACCESS EXCLUSIVE MODE",
	)); err != nil {
		return false, err
	}
	live, err := readPostgresHistory(ctx, tx, namespace)
	if err != nil {
		return false, err
	}
	if err := ValidateHistory(Postgres, live); err != nil {
		return false, err
	}
	current := 0
	if len(live) != 0 {
		current = live[len(live)-1].Version
	}
	if direction == Up {
		if current >= intended.Migration.Version {
			return false, tx.Commit(ctx)
		}
		if current+1 != intended.Migration.Version {
			return false, &HistoryError{Message: fmt.Sprintf(
				"concurrent migration moved current version to %d; expected %d",
				current, intended.Migration.Version-1,
			)}
		}
		if err := execPostgresBatch(ctx, tx, namespace.Render(intended.Migration.UpSQL)); err != nil {
			return false, err
		}
		// requiredTables/columns describe latest, not each intermediate version. A
		// fresh install must be allowed to commit v1 before v2 creates its objects.
		if intended.Migration.Version == LatestVersion(Postgres) {
			shape, err := missingPostgresSchema(ctx, tx, namespace)
			if err != nil {
				return false, err
			}
			if len(shape) != 0 {
				return false, &SchemaError{Messages: shape}
			}
		}
		if _, err := tx.Exec(ctx, namespace.Render(`
INSERT INTO headgate_schema_migration (line, version, name, checksum, applied_at_ms)
VALUES ('main', $1, $2, $3,
        (EXTRACT(EPOCH FROM clock_timestamp()) * 1000)::bigint)`),
			int64(intended.Migration.Version), intended.Migration.Name, Checksum(intended.Migration)); err != nil {
			return false, err
		}
	} else {
		if current < intended.Migration.Version {
			return false, tx.Commit(ctx)
		}
		if current != intended.Migration.Version {
			return false, &HistoryError{Message: fmt.Sprintf(
				"cannot migrate down version %d; current version is %d", intended.Migration.Version, current,
			)}
		}
		if err := execPostgresBatch(ctx, tx, namespace.Render(intended.Migration.DownSQL)); err != nil {
			return false, err
		}
		if _, err := tx.Exec(ctx, namespace.Render(
			`DELETE FROM headgate_schema_migration WHERE line = 'main' AND version = $1`,
		), int64(intended.Migration.Version)); err != nil {
			return false, err
		}
	}
	if err := tx.Commit(ctx); err != nil {
		return false, err
	}
	return true, nil
}

func execPostgresBatch(ctx context.Context, tx pgx.Tx, sql string) error {
	_, err := tx.Conn().PgConn().Exec(ctx, sql).ReadAll()
	return err
}

func adoptPostgresScoped(
	ctx context.Context,
	conn *pgx.Conn,
	namespace postgressql.Namespace,
) ([]AppliedMigration, error) {
	state, existing, err := appliedPostgresScoped(ctx, conn, namespace)
	if err != nil {
		return nil, err
	}
	if state == Versioned {
		return existing, ValidateHistory(Postgres, existing)
	}
	if state == Empty {
		return nil, fmt.Errorf("cannot adopt an empty database; migrate up instead")
	}
	if _, err := conn.Exec(ctx, namespace.Render(createPostgresHistory)); err != nil {
		return nil, err
	}
	tx, err := conn.Begin(ctx)
	if err != nil {
		return nil, err
	}
	defer func() { _ = tx.Rollback(ctx) }()
	if _, err := tx.Exec(ctx, namespace.Render(
		"LOCK TABLE headgate_schema_migration IN ACCESS EXCLUSIVE MODE",
	)); err != nil {
		return nil, err
	}
	live, err := readPostgresHistory(ctx, tx, namespace)
	if err != nil {
		return nil, err
	}
	if len(live) != 0 {
		if err := ValidateHistory(Postgres, live); err != nil {
			return nil, err
		}
		if err := tx.Commit(ctx); err != nil {
			return nil, err
		}
		return live, nil
	}
	shape, err := missingPostgresSchema(ctx, tx, namespace)
	if err != nil {
		return nil, err
	}
	if len(shape) != 0 {
		return nil, &SchemaError{Messages: shape}
	}
	insert := namespace.Render(`
INSERT INTO headgate_schema_migration (line, version, name, checksum, applied_at_ms)
VALUES ('main', $1, $2, $3,
        (EXTRACT(EPOCH FROM clock_timestamp()) * 1000)::bigint)`)
	for _, migration := range byBackend[Postgres] {
		if _, err := tx.Exec(ctx, insert,
			int64(migration.Version), migration.Name, Checksum(migration)); err != nil {
			return nil, err
		}
	}
	if err := tx.Commit(ctx); err != nil {
		return nil, err
	}
	_, adopted, err := appliedPostgresScoped(ctx, conn, namespace)
	return adopted, err
}

func AdoptPostgres(ctx context.Context, conn *pgx.Conn) ([]AppliedMigration, error) {
	namespace, err := postgresNamespace(ctx, conn, nil)
	if err != nil {
		return nil, err
	}
	return adoptPostgresScoped(ctx, conn, namespace)
}

func AdoptPostgresInSchema(
	ctx context.Context,
	conn *pgx.Conn,
	schema string,
) ([]AppliedMigration, error) {
	namespace, err := postgresNamespace(ctx, conn, &schema)
	if err != nil {
		return nil, err
	}
	return adoptPostgresScoped(ctx, conn, namespace)
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
